legend-engine-ide-lsp-server/src/main/java/org/finos/legend/engine/ide/lsp/server/ExtensionsGuard.java

```diff
@@ -0,0 +1,118 @@
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package org.finos.legend.engine.ide.lsp.server;
+
+import java.io.Closeable;
+import java.util.function.Supplier;
+import org.finos.legend.engine.ide.lsp.extension.LegendLSPGrammarExtension;
+import org.finos.legend.engine.ide.lsp.extension.LegendLSPGrammarLibrary;
+import org.finos.legend.engine.ide.lsp.extension.LegendLSPInlineDSLExtension;
+import org.finos.legend.engine.ide.lsp.extension.LegendLSPInlineDSLLibrary;
+
+class ExtensionsGuard
+{
+    private final Iterable<LegendLSPGrammarExtension> providedGrammarExtensions;
+    private final Iterable<LegendLSPInlineDSLExtension> providedInlineDSLs;
+    private final LegendLanguageServer server;
+    private volatile ClassLoader classLoader;
+    private volatile LegendLSPGrammarLibrary grammars;
+    private volatile LegendLSPInlineDSLLibrary inlineDSLs;
+
+    public ExtensionsGuard(LegendLanguageServer server, LegendLSPGrammarLibrary providedGrammarExtensions, LegendLSPInlineDSLLibrary providedInlineDSLs)
+    {
+        this.server = server;
+        this.grammars = providedGrammarExtensions;
+        this.inlineDSLs = providedInlineDSLs;
+        this.providedGrammarExtensions = providedGrammarExtensions.getExtensions();
+        this.providedInlineDSLs = providedInlineDSLs.getExtensions();
+    }
+
+    public synchronized void initialize(ClassLoader classLoader)
+    {
+        this.server.logInfoToClient("Initializing grammar extensions");
+
+        ClassLoader contextClassLoader = Thread.currentThread().getContextClassLoader();
+
+        try
+        {
+            if (this.classLoader != null && this.classLoader instanceof Closeable)
+            {
+                ((Closeable) this.classLoader).close();
+            }
+
+            Thread.currentThread().setContextClassLoader(classLoader);
+
+            this.grammars = LegendLSPGrammarLibrary.builder().withExtensions(this.providedGrammarExtensions).withExtensionsFrom(classLoader).build();
+            this.inlineDSLs = LegendLSPInlineDSLLibrary.builder().withExtensions(this.providedInlineDSLs).withExtensionsFrom(classLoader).build();
+
+            this.server.logInfoToClient("Grammar extensions available: " + String.join(", ", this.grammars.getExtensionNames()));
+            this.server.logInfoToClient("Inline DSL extensions available: " + String.join(", ", this.inlineDSLs.getExtensionNames()));
+
+            this.classLoader = classLoader;
+        }
+        catch (Throwable e)
+        {
+            throw new RuntimeException(e);
+        }
```

Comment:
Is this catch needed? Is it just for closing the possible old class loader? If so, let's just catch IOException and wrap with UncheckedIOException. If not, let's try to avoid unnecessary wrapping of RuntimeExceptions.